### PR TITLE
Move BandwidthAllocator and BitrateControllerPacketHandler to kotlin

### PR DESCRIPTION
**Pull request by bgrozev**

### CODE REVIEW

File: jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/PacketHandler.kt [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+     * a source.
+     */
+    private fun lookupOrCreateAdaptiveSourceProjection(singleAllocation: SingleAllocation): AdaptiveSourceProjection? {
+        val source = singleAllocation.mediaSource
+        val endpointID = singleAllocation.endpointId
+        if (source == null) {
+            return null
+        }
+        synchronized(adaptiveSourceProjectionMap) {
+            var adaptiveSourceProjection = adaptiveSourceProjectionMap[source.primarySSRC]
```
Comment (JonathanLennox) [inline jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/PacketHandler.kt:150]: Using `adaptiveSourceProjection` as a `var` both for the result of the lookup and for the newly-constructed object doesn't seem very Kotlin-y, but I'm not sure what a better idiom would be.  Maybe have the result of the lookup be `val existingProjection` or something?

File: jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/PacketHandler.kt [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+            if (ArrayUtils.isNullOrEmpty(rtpEncodings)) {
+                return null
+            }
+
+            // XXX the lambda keeps a reference to the sourceBitrateAllocation
+            // (a short lived object under normal circumstances) which keeps
+            // a reference to the Endpoint object that it refers to. That
+            // can cause excessive object retention (i.e. the endpoint is expired
+            // but a reference persists in the adaptiveSourceProjectionMap). We're
+            // creating local final variables and pass that to the lambda function
```
Comment (JonathanLennox) [inline jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/PacketHandler.kt:164]: "final variables" is the Java idiom, in Kotlin they're `val`s.

Also the grammar of this comment is a bit wonky, though that's not new from the Java version.

File: jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/PacketHandler.kt [hunk 3]
```diff
@@ -166,7 +168,7 @@ internal class PacketHandler(
         }
     }
 
-    fun timeSinceFirstMedia(): Long = if (firstMediaMs == -1L) -1 else clock.instant().toEpochMilli() - firstMediaMs
+    fun timeSinceFirstMedia(): Duration = firstMedia?.let { Duration.between(it, clock.instant()) } ?: Duration.ZERO
```
Comment (JonathanLennox) [inline jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/PacketHandler.kt:186]: I'm not crazy about using `Duration.ZERO` as "never", though I guess for how we're using it it doesn't matter.

File: jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/BandwidthAllocator.kt [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+internal class BandwidthAllocator<T : MediaSourceContainer>(
+    eventHandler: EventHandler,
+    /**
+     * Provide the current list of endpoints (in no particular order).
+     * TODO: Simplify to avoid the weird (and slow) flow involving `endpointsSupplier` and `sortedEndpointIds`.
+     */
+    private val endpointsSupplier: Supplier<List<T>>,
+    /**
+     * Whether bandwidth allocation should be constrained to the available bandwidth (when `true`), or assume
+     * infinite bandwidth (when `false`.
```
Comment (JonathanLennox) [inline jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/BandwidthAllocator.kt:45]: Missing close paren

File: jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/BandwidthAllocator.kt [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+
+    /**
+     * The task scheduled to call [.update].
+     */
+    private var updateTask: ScheduledFuture<*>? = null
+
+    init {
+        eventEmitter.addHandler(eventHandler)
+        // Don't trigger an update immediately, the settings might not have been configured.
+        lastUpdateTime = clock.instant()
```
Comment (JonathanLennox) [inline jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/BandwidthAllocator.kt:107]: Is this in `init` to make sure it happens after `addHandler` and before `rescheduleUpdate`?

File: jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/BandwidthAllocator.kt [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+        val suspendedIds = sourceBitrateAllocations
+            .filter { it.isSuspended }
+            .mapNotNull { it.mediaSource.sourceName }.toList()
+        if (suspendedIds.isNotEmpty()) {
+            logger.info("Sources suspended due to insufficient bandwidth (bwe=$availableBandwidth bps): $suspendedIds")
+        }
+        val allocations: MutableSet<SingleAllocation> = HashSet()
+        var targetBps: Long = 0
+        var idealBps: Long = 0
+        for (sourceBitrateAllocation: SingleSourceAllocation in sourceBitrateAllocations) {
```
Comment (JonathanLennox) [inline jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/BandwidthAllocator.kt:276]: Does the type need to be specified here?

File: jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/BandwidthAllocator.kt [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+        return BandwidthAllocation(allocations, oversending, idealBps, targetBps, suspendedIds)
+    }
+
+    /**
+     * Query whether this allocator is forwarding a source from a given endpoint, as of its
+     * most recent allocation decision.
+     */
+    fun isForwarding(endpointId: String): Boolean {
+        return allocation.isForwarding(endpointId)
+    }
```
Comment (JonathanLennox) [inline jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/BandwidthAllocator.kt:290]: I might use Kotlin `=` syntax for a one-liner like this.

File: jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/BandwidthAllocator.kt [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+     * Also, re-schedule the next update in at most `maxTimeBetweenCalculations`. This should only be run
+     * in the constructor or in the scheduler thread, otherwise it will schedule multiple tasks.
+     */
+    private fun rescheduleUpdate() {
+        if (expired) {
+            return
+        }
+        val timeSinceLastUpdate = Duration.between(lastUpdateTime, clock.instant())
+        val period = BitrateControllerConfig.config.maxTimeBetweenCalculations()
+        val delayMs: Long = if (timeSinceLastUpdate > period) {
```
Comment (JonathanLennox) [inline jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/BandwidthAllocator.kt:342]: Does this need a type specified?

File: jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/BandwidthAllocator.kt (hunk 7 above)
Comment (bgrozev) [reply jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/BandwidthAllocator.kt]: I removed the function, it wasn't used

File: jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/BandwidthAllocator.kt (hunk 5 above)
Comment (bgrozev) [reply jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/BandwidthAllocator.kt]: No, that's just how IDEA translated it to kt. I cleaned it up

File: jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/PacketHandler.kt (hunk 2 above)
Comment (bgrozev) [reply jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/PacketHandler.kt]: The comment and use of a final variable is no longer valid (we only pass a MediaSourceDesc which doesn't have back references to Endpoint), so I removed it.

File: jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/PacketHandler.kt (hunk 1 above)
Comment (bgrozev) [reply jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/PacketHandler.kt]: Cleaned up

**Review by JonathanLennox — APPROVED** [review]

**Pull request merged**